frontend/i_types.py

```diff
@@ -0,0 +1,274 @@
+from abc import ABCMeta, abstractmethod
+
+class Type(metaclass=ABCMeta):
+    """This class is the top-parent of all possible inferred types of a Python program.
+    Every type has its own class that inherits this class.
+    """
+
+    @abstractmethod
+    def is_subtype(self, t):
+        """Check if this type is a subtype of the parameter type.
+
+        Arguments:
+            type (Type): The type to check the subtype relationship against.
+        """
+        pass
+
+
+    @abstractmethod
+    def get_name(self):
+        """Get the name of the type"""
+        pass
+
+    # TODO: Implement __eq__ and __hash__ for all types, for correct set behaviour
+
+
+class TNone(Type):
+
+    def is_subtype(self, t):
+        return True
+
+    def get_name(self):
+        return "None"
+
+
+class TBool(Type):
+
+    def is_subtype(self, t):
+        return isinstance(t, (TBool, TInt, TFloat))
+
+    def get_name(self):
+        return "Bool"
+
+
+class TInt(Type):
+
+    def is_subtype(self, t):
+        return isinstance(t, (TInt, TFloat))
+	
+    def get_name(self):
+        return "Int"
+
+
+class TFloat(Type):
+
+    def is_subtype(self, t):
+        return isinstance(t, TFloat)
+
+    def get_name(self):
+        return "Float"
+
+
+class TString(Type):
+
+    def is_subtype(self, t):
+        return isinstance(t, TString)
+
+    def get_name(self):
+        return "String"
+
+
+class TBytesString(Type):
+
+    def is_subtype(self, t):
+        return isinstance(t, TBytesString)
+
+    def get_name(self):
+        return "ByteString"
+
+
+class TList(Type):
+    """Type given to homogeneous lists.
+
+    Attributes:
+        type (Type): Type of the list elements
+    """
+
+    def __init__(self, t):
+        self.type = t
+
+    def is_subtype(self, t):
+        return isinstance(t, TList) and self.type.is_subtype(t.type)
```

Comment:
> you would want the type argument in both types to be identical.

What about the relationship between `List[int]` and `List[float]`? I think the first should be a subtype of the latter. Also why shouldn't `List[str]` be a subtype of `List[object]`. If a function, for example, returns both `List[str]` and `List[object]`, I can see that it is more appropriate to have only `List[object]` as the only return type. Please explain more why do you think so.